### PR TITLE
Bug 1232276 - Repeat sync if needed, r=ferjm

### PR DESCRIPTION
**Pull request by michielbdejong**
This anticipates the new `fetchToken` option and `SyncResults.fetchToken` value in Kinto.js, see https://github.com/Kinto/kinto.js/pull/299

### CODE REVIEW

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=b1c8b47a9a77dd80aa92504f7ab75bc2ace00f5a

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=21d334b1e057ddce819c88af1f188edcfd36d77f

**Comment by michielbdejong** [comment]
Unit test added.

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=4940545dd3a0727f59dde1c8a9875e8a8aa77eb4

File: apps/sharedtest/test/unit/sync/sync-engine/syncengine_test.js [hunk 1]
```diff
@@ -288,10 +288,40 @@ ld be a Function`);
...
+            reset: function() { }
+          }
+        }
+      };
+      var se = new SyncEngine(credentials);
+      for (var i = 0; i < 8; i++) {
+        FxSyncWebCrypto.responses.push({ foo: i });
+      }
+      se.syncNow({ 'bookmarks': {
+        limit: 3
```
Comment (michielbdejong) [inline apps/sharedtest/test/unit/sync/sync-engine/syncengine_test.js]: Note that `limit` will be renamed to `fetchLimit` in https://github.com/mozilla-b2g/gaia/pull/33745/files#diff-795a9307cc60d72cc4ceaa37fddaf7b5R279

File: shared/js/sync/engine.js [hunk 2]
```diff
@@ -401,9 +409,19 @@ ccount`);
...
+          if (syncResults2.fetchToken) {
+            collectionOptions.fetchToken = syncResults2.fetchToken;
+            repeatNeeded = true;
+          } else {
+            repeatNeeded = false;
+          }
           this._haveUnsyncedConflicts[collectionName] = false;
         });
+      }).then(() => {
+        if (repeatNeeded) {
```
Comment (ferjm) [inline shared/js/sync/engine.js]: Could we add a pref to enable/disable repetition, please? For the phone we don't want to fetch the whole history for now. At least until https://bugzilla.mozilla.org/show_bug.cgi?id=1231933 is fixed.

File: shared/js/sync/engine.js [hunk 3]
```diff
@@ -387,13 +390,18 @@ payload as JSON`);
     },
 
     _updateCollection: function(collectionName, collectionOptions) {
```
Comment (ferjm) [inline shared/js/sync/engine.js]: We could use some good comments for this method :)

File: shared/js/sync/engine.js (hunk 2 above)
Comment (michielbdejong) [inline shared/js/sync/engine.js]: OK - if we don't want to merge this for the TV and also not for the phone then let's just shelf it. If at some point we want to enable it for one of the two but not for the other, I'll add the pref for that.

**Comment by michielbdejong** [comment]
Cleaning up PRs that are > 1 year old.

**Pull request closed without merging**